File: javascript/55.jump-game.js
/*
 * @lc app=leetcode id=55 lang=javascript
 *
 * [55] Jump Game
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
  // recursively iterate over all jumps from index
  // O(n^2) time
  // O(n^2) space
  let deadIndices = new Set()
  let canJumpHelper = (currIndex) => {
    if (deadIndices.has(currIndex)) return false
    if (currIndex + nums[currIndex] >= nums.length -1) return true
    if (currIndex === nums.length -1) return true
    if (currIndex > nums.length -1) return false

    for (let i = nums[currIndex]; i > 0; i--) {
      if (canJumpHelper(currIndex + i)) return true
    }
    deadIndices.add(currIndex)
    return false
  }
  return canJumpHelper(0)
};
// @lc code=end

File: cpp/55.jump-game.cpp
/*
 * @lc app=leetcode id=55 lang=cpp
 *
 * [55] Jump Game
 */

// @lc code=start
class Solution
{
private:
    unordered_set<int> numCache;

public:
    bool dfsHelper(vector<int> &nums, int pos)
    {
        // base cases
        if (pos == nums.size() - 1)
        {
            return true;
        }
        else if (pos + nums[pos] >= nums.size() - 1)
        {
            return true;
        }
        else if (nums[pos] < 1)
        {
            return false;
        }
        else if (pos >= nums.size())
        {
            return false;
        }
        else if (numCache.count(pos) > 0)
        {
            return false;
        }
        // recursive subcalls
        for (int i = nums[pos]; i > 0; i--)
        {
            if (dfsHelper(nums, pos + i))
            {
                return true;
            }
        }
        // cache subproblems
        numCache.insert(pos);
        return false;
    }
    bool canJump(vector<int> &nums)
    {
        // DFS recursive approach
        // at every index iterate the number of steps
        // until it becomes clear you can reach end
        // cache subproblems
        // O(n^2(averageNumOfSteps)) time O(n^(averageNumOfSteps)) space
        auto res = dfsHelper(nums, 0);
        return res;
    }
};
// @lc code=end
